music_assistant/server/providers/snapcast/__init__.py

```diff
@@ -0,0 +1,305 @@
+        ffmpeg = (
+            FFmpeg()
+            .option("y")
+            .input(url)
+            .output(f"tcp://{stream_host}", f="u16le", acodec="pcm_s16le", ac=2, ar=48000)
+        )
+        self.mass.create_task(ffmpeg.execute())
+
+        @ffmpeg.on("start")
+        async def on_start(arguments: list[str]):
+            self.logger.debug("Ffmpeg stream is running")
+            if hasattr(stream, "ffmpeg"):
+                await self.cmd_stop(player_id)
+            stream.ffmpeg = ffmpeg
+            player.state = PlayerState.PLAYING
+            player.current_url = url
+            player.elapsed_time = 0
+            player.elapsed_time_last_updated = time.time()
+            self.mass.players.register_or_update(player)
+
+        @ffmpeg.on("progress")
+        def on_progress(progress: Progress):
+            player.state = PlayerState.PLAYING
+            self.mass.players.register_or_update(player)
+
+        @ffmpeg.on("completed")
+        async def on_completed():
+            player.state = PlayerState.IDLE
+            self.mass.players.register_or_update(player)
+
+        @ffmpeg.on("terminated")
+        async def on_terminated():
+            player.state = PlayerState.IDLE
+            self.mass.players.register_or_update(player)
+
+    async def cmd_stop(self, player_id: str) -> None:
+        """Send STOP command to given player."""
+        player = self.mass.players.get(player_id, raise_unavailable=False)
+        if player.state != PlayerState.IDLE:
+            stream = self._get_snapstream(player_id)
+            if hasattr(stream, "ffmpeg"):
+                try:
+                    stream.ffmpeg.terminate()
+                    self.logger.debug("ffmpeg player stopped")
+                except FFmpegError:
+                    self.logger.debug("Fail to stop ffmpeg player")
+
+    async def cmd_pause(self, player_id: str) -> None:
+        """Send PAUSE command to given player."""
+        await self.cmd_stop(player_id)
+
+    async def cmd_volume_mute(self, player_id, muted):
+        """Send MUTE command to given player."""
+        self.mass.create_task(self._snapserver.client(player_id).set_muted(muted))
+
+    async def _remove_stream(self, stream_id):
+        self.mass.create_task(self._server.stream_remove_stream(stream_id))
+
+    def _snapclient_get_group_clients_identifiers(self, player_id):
+        group = self._get_snapgroup(player_id)
+        return [ele for ele in group.clients if ele != player_id]
+
+    async def cmd_sync(self, player_id: str, target_player: str) -> None:
+        """Sync Snapcast player."""
+        child_player = self.mass.players.get(player_id)
+        assert child_player  # guard
+        parent_player = self.mass.players.get(target_player)
+        assert parent_player  # guard
+        # always make sure that the parent player is part of the sync group
+        # parent_player.group_childs.add(parent_player.player_id)
```

Comment:
is this leftover ?

Reply:
Yes, I remove it

Reply:
I removed it and changed the code so that these events are also processed through a callback.